### PR TITLE
Add ROCm HIP library

### PR DESCRIPTION
**Pull request by cgmb**
This change adds the HIP runtime for the AMD GPU platform (ROCm) to the library install list.

I had to specify the untar directory to make this work, but it doesn't seem like other libraries needed to do that so I wonder if perhaps I screwed up the tar command in https://github.com/compiler-explorer/misc-builder/pull/47.

~I marked `install_always: true` for the hip-runtime-amd in the CUDA section, as this library is necessary to properly build anything with HIP for AMDGPUs. I figured that if the HIP compilers are used in the CUDA section, then to ensure they work when the language is set to CUDA, the runtime library should always be installed when the language is CUDA. However,~ my understanding of the overall Compiler Explorer system is very shallow and it's possible that I'm approaching this in a suboptimal way.

### CODE REVIEW

**Comment by cgmb** [comment]
I opened this as a draft because it works, but I feel like it might be better to change the tar command in misc builder rather than setting the `untar_dir` here.

File: bin/yaml/libraries.yaml [hunk 1]
```diff
@@ -1658,6 +1670,19 @@ libraries:
...
+      path_name: libs/rocm/{name}
+      s3_path_prefix: hip-amd-rocm-{name}
+      subdir: libs/rocm/{name}
+      untar_dir: opt/compiler-explorer/libs/rocm/{{name}}
+      targets:
+      - 4.5.2
+      - 5.0.2
+      - 5.1.3
+      - 5.2.3
+      install_always: true
```
Comment (jrhemstad) [inline bin/yaml/libraries.yaml:1684]: @mattgodbolt or @partouf can you remind me what the side-effects of "install_always" are again? 

Hopefully this doesn't mean it automatically links against this library by default as that would be undesirable.

Comment (partouf) [reply bin/yaml/libraries.yaml:1684]: install_always means it's always installed even if we already have it. That's generally a bad idea unless it's a trunk version (but all trunk/nightly things are automatically tagged as install_always)
So it really shouldn't be there.

Comment (cgmb) [reply bin/yaml/libraries.yaml:1684]: That's different from what I thought it did. Thanks for the correction!

**Comment by cgmb** [comment]
I've opened https://github.com/compiler-explorer/misc-builder/pull/48 to fix the tarballs on S3 and I've updated to match the corrected paths.

**Comment by mattgodbolt** [comment]
I'll build those updated rocms now

**Review by mattgodbolt — CHANGES_REQUESTED** [review]

File: bin/yaml/libraries.yaml [hunk 2]
```diff
@@ -1658,6 +1670,18 @@ libraries:
       targets:
       - 1.8.0
       type: github
+    hip-amd:
+      check_file: include/hip/hip_runtime.h
+      path_name: libs/rocm/{name}
+      s3_path_prefix: hip-amd-rocm-{name}
+      subdir: libs/rocm/{name}
```
[the remark this answers is not in the document]
Comment (mattgodbolt) [reply bin/yaml/libraries.yaml:1677]: It's possible you rooted this under each language separately but in this case they're shared I think, so... the source and destination are the same in the two cases (right? It's been a long day so I may have missed something)

[the remark this answers is not in the document]
Comment (cgmb) [reply bin/yaml/libraries.yaml:1677]: One is under the C++ section, one is under the CUDA section. The headers can be used from both languages. I put it in both for that reason, but if it should just be in one then we can remove it from the other.

**Comment by mattgodbolt** [comment]
* https://github.com/compiler-explorer/infra/actions/runs/3222892165
* https://github.com/compiler-explorer/infra/actions/runs/3222894345
* https://github.com/compiler-explorer/infra/actions/runs/3222894721
* https://github.com/compiler-explorer/infra/actions/runs/3222894918

File: bin/yaml/libraries.yaml (hunk 2 above)
[the remark this answers is not in the document]
Comment (mattgodbolt) [reply bin/yaml/libraries.yaml:1677]: Got it! yes that's confusing: the headers here are for organisation really only - it doesn't even have to be under a language at all. For the easiest let's leave it under C++ - duplicates will make some things more tricky going forward (and I should have a check for duplicate destinations in rules, that's on me).

[the remark this answers is not in the document]
Comment (mattgodbolt) [reply bin/yaml/libraries.yaml:1677]: (sorry for this not being clearer, this is our first real cross-platform library; though we have some C libraries that are usable in C++, they're only listed in one place etc)

[the remark this answers is not in the document]
Comment (cgmb) [reply bin/yaml/libraries.yaml]: That makes sense. I've made the change and they can now install with the new tarballs on S3.

**Review by mattgodbolt — APPROVED** [review]
LGTM! Will merge and install

**Comment by mattgodbolt** [comment]
```
$ sudo ce_install install hip-amd
2022-10-11 01:53:58,371 lib.installation_context INFO     Making uncached requests
2022-10-11 01:53:58,372 lib.ce_install  INFO     Creating multiprocessing pool with 2 workers
Installing libraries/c++/hip-amd 4.5.2
2022-10-11 01:54:02,264 lib.installation_context INFO     Fetching https://s3.amazonaws.com/compiler-explorer/opt/hip-amd-rocm-4.5.2.tar.xz (37156500 bytes)
2022-10-11 01:54:03,328 lib.installation_context INFO     100% of https://s3.amazonaws.com/compiler-explorer/opt/hip-amd-rocm-4.5.2.tar.xz
2022-10-11 01:54:03,329 lib.installation_context INFO     Piping to tar Jxf -
2022-10-11 01:54:12,071 lib.installation_context INFO     Moving from staging (/opt/compiler-explorer/staging/f96664cb-de30-4566-97d9-51b4dfe84485/hip-amd-rocm-4.5.2) to final destination (/opt/compiler-explorer/libs/rocm/4.5.2)
2022-10-11 01:54:12,668 lib.installation_context INFO     Destination /opt/compiler-explorer/libs/rocm/4.5.2 exists, temporarily moving out of the way (to /opt/compiler-explorer/staging/f96664cb-de30-4566-97d9-51b4dfe84485/temp_orig)
2022-10-11 01:54:12,708 lib.ce_install  INFO     libraries/c++/hip-amd 4.5.2 installed OK
Installing libraries/c++/hip-amd 5.0.2
2022-10-11 01:54:12,932 lib.installation_context INFO     Fetching https://s3.amazonaws.com/compiler-explorer/opt/hip-amd-rocm-5.0.2.tar.xz (38887680 bytes)
2022-10-11 01:54:13,603 lib.installation_context INFO     100% of https://s3.amazonaws.com/compiler-explorer/opt/hip-amd-rocm-5.0.2.tar.xz
2022-10-11 01:54:13,604 lib.installation_context INFO     Piping to tar Jxf -
2022-10-11 01:54:45,305 lib.installation_context INFO     Moving from staging (/opt/compiler-explorer/staging/63f308ec-db42-452d-bee8-35b6d6b9bd62/hip-amd-rocm-5.0.2) to final destination (/opt/compiler-explorer/libs/rocm/5.0.2)
2022-10-11 01:54:45,909 lib.installation_context INFO     Destination /opt/compiler-explorer/libs/rocm/5.0.2 exists, temporarily moving out of the way (to /opt/compiler-explorer/staging/63f308ec-db42-452d-bee8-35b6d6b9bd62/temp_orig)
2022-10-11 01:54:45,952 lib.ce_install  INFO     libraries/c++/hip-amd 5.0.2 installed OK
Installing libraries/c++/hip-amd 5.1.3
2022-10-11 01:54:46,339 lib.installation_context INFO     Fetching https://s3.amazonaws.com/compiler-explorer/opt/hip-amd-rocm-5.1.3.tar.xz (39302132 bytes)
2022-10-11 01:54:47,068 lib.installation_context INFO     100% of https://s3.amazonaws.com/compiler-explorer/opt/hip-amd-rocm-5.1.3.tar.xz
2022-10-11 01:54:47,069 lib.installation_context INFO     Piping to tar Jxf -
2022-10-11 01:55:01,252 lib.installation_context INFO     Moving from staging (/opt/compiler-explorer/staging/2f90b981-fdc0-46b6-8cf9-7ea9e3fecc36/hip-amd-rocm-5.1.3) to final destination (/opt/compiler-explorer/libs/rocm/5.1.3)
2022-10-11 01:55:02,970 lib.installation_context INFO     Destination /opt/compiler-explorer/libs/rocm/5.1.3 exists, temporarily moving out of the way (to /opt/compiler-explorer/staging/2f90b981-fdc0-46b6-8cf9-7ea9e3fecc36/temp_orig)
2022-10-11 01:55:03,014 lib.ce_install  INFO     libraries/c++/hip-amd 5.1.3 installed OK
Installing libraries/c++/hip-amd 5.2.3
2022-10-11 01:55:03,227 lib.installation_context INFO     Fetching https://s3.amazonaws.com/compiler-explorer/opt/hip-amd-rocm-5.2.3.tar.xz (39722808 bytes)
2022-10-11 01:55:04,097 lib.installation_context INFO     100% of https://s3.amazonaws.com/compiler-explorer/opt/hip-amd-rocm-5.2.3.tar.xz
2022-10-11 01:55:04,098 lib.installation_context INFO     Piping to tar Jxf -
2022-10-11 01:55:21,721 lib.installation_context INFO     Moving from staging (/opt/compiler-explorer/staging/62f57e95-6b08-41e5-99b0-a33f04181fa0/hip-amd-rocm-5.2.3) to final destination (/opt/compiler-explorer/libs/rocm/5.2.3)
2022-10-11 01:55:23,816 lib.installation_context INFO     Destination /opt/compiler-explorer/libs/rocm/5.2.3 exists, temporarily moving out of the way (to /opt/compiler-explorer/staging/62f57e95-6b08-41e5-99b0-a33f04181fa0/temp_orig)
2022-10-11 01:55:23,858 lib.ce_install  INFO     libraries/c++/hip-amd 5.2.3 installed OK
4 packages installed OK, 0 skipped, and 0 failed installation
```

**Comment by mattgodbolt** [comment]
```
Oct/11 01:55 admin-node~/infra (main|✔) 1.4m $ ls /opt/compiler-explorer/libs/rocm
4.5.2  5.0.2  5.1.3  5.2.3
```
looks good!

**Pull request merged**